src/com/xilinx/rapidwright/eco/ECOTools.java

```diff
@@ -0,0 +1,1021 @@
+
+            String[] pins = net_pins[1].split("[{} ]+");
+            List<EDIFHierPortInst> portInsts = netPortInsts.computeIfAbsent(ehn, (n) -> new ArrayList<>(pins.length));
+            for (String pin : pins) {
+                if (pin.isEmpty()) {
+                    continue;
+                }
+
+                EDIFHierPortInst ehpi = netlist.getHierPortInstFromName(pin);
+                if (ehpi == null) {
+                    // No pin exists currently; create and attach one to its port
+                    int pos = pin.lastIndexOf(EDIFTools.EDIF_HIER_SEP);
+                    String path = pin.substring(0, pos);
+                    EDIFHierCellInst ehci = netlist.getHierCellInstFromName(path);
+                    if (ehci == null) {
+                        throw new RuntimeException("ERROR: Unable to find inst '" + path + "' corresponding to pin '" + pin + "'");
+                    }
+                    String name = pin.substring(pos+1);
+                    EDIFCell cell = ehci.getCellType();
+                    EDIFPort ep = cell.getPort(name);
+                    if (ep == null) {
+                        ep = cell.getPort(EDIFTools.getRootBusName(name, false));
+                    }
+                    if (ep == null) {
+                        throw new RuntimeException("ERROR: Unable to find port '" + name + "' on inst '" + path + "' corresponding to pin '" + pin + "'");
+                    }
+                    EDIFPortInst epi;
+                    if (ep.isBus()) {
+                        int portIdx = EDIFTools.getPortIndexFromName(name);
+                        if (ep.isLittleEndian()) {
+                            portIdx = ep.getWidth() - 1 - portIdx;
+                        }
+                        epi = new EDIFPortInst(ep, null, portIdx, ehci.getInst());
+                    } else {
+                        epi = new EDIFPortInst(ep, null, ehci.getInst());
+                    }
+                    ehpi = new EDIFHierPortInst(ehci.getParent(), epi);
+                }
+                portInsts.add(ehpi);
+            }
+        }
+        connectNet(design, netPortInsts, deferredRemovals);
+    }
+
+    /**
+     * Creates a new SitePinInst source for the physical net provided.  It will route out an
+     * internal pin to a site pin output.
+     * @param design The current design
+     * @param targetHierSrc The hierarchical source pin
+     * @param newPhysNet The net to add the new source pin to
+     * @return The newly created source site pin
+     */
+    private static SitePinInst routeOutSitePinInstSource(Design design,
+                                                         EDIFHierPortInst targetHierSrc,
+                                                         Net newPhysNet) {
+        // This net is internal to a site, need to create a site pin and route out to it
+        Cell srcCell = design.getCell(targetHierSrc.getFullHierarchicalInstName());
```

Comment:
Is it worth adding an assert here to ensure `targetHierSrc` is indeed a source pin?